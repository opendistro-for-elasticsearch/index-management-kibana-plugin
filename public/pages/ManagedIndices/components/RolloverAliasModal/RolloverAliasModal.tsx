/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { Component } from "react";
import {
  EuiButton,
  EuiButtonEmpty,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiOverlayMask,
  EuiFormRow,
  EuiFieldText,
} from "@elastic/eui";
import { BrowserServices } from "../../../../models/interfaces";
import { getErrorMessage } from "../../../../utils/helpers";
import { CoreStart } from "kibana/public";

interface RolloverAliasModalProps {
  onClose: () => void;
  services: BrowserServices;
  index: string;
  core: CoreStart;
}

interface RolloverAliasModalState {
  rolloverAlias: string;
}

export default class RolloverAliasModal extends Component<RolloverAliasModalProps, RolloverAliasModalState> {
  state: RolloverAliasModalState = {
    rolloverAlias: "",
  };

  onEditRolloverAlias = async (): Promise<void> => {
    const {
      onClose,
      index,
      services: { indexService },
    } = this.props;
    const { rolloverAlias } = this.state;
    try {
      const response = await indexService.editRolloverAlias(index, rolloverAlias);
      if (response.ok) {
        if (response.response.acknowledged) {
          this.props.core.notifications.toasts.addSuccess(`Edited rollover alias on ${index}`);
        } else {
          this.props.core.notifications.toasts.addDanger(`Failed to edit rollover alias on ${index}`);
        }
      } else {
        this.props.core.notifications.toasts.addDanger(response.error);
      }
      onClose();
    } catch (err) {
      this.props.core.notifications.toasts.addDanger(getErrorMessage(err, `There was a problem editing rollover alias on ${index}`));
    }
  };

  onChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({ rolloverAlias: e.target.value });
  };

  render() {
    const { rolloverAlias } = this.state;
    const { onClose } = this.props;
    return (
      <EuiOverlayMask>
        {/*
            // @ts-ignore */}
        <EuiModal onCancel={onClose} onClose={onClose}>
          <EuiModalHeader>
            <EuiModalHeaderTitle>Edit rollover alias</EuiModalHeaderTitle>
          </EuiModalHeader>

          <EuiModalBody>
            <EuiFormRow label="Rollover alias" helpText="A rollover alias is required when using the rollover action.">
              <EuiFieldText placeholder="Rollover alias" value={rolloverAlias} onChange={this.onChange} />
            </EuiFormRow>
          </EuiModalBody>

          <EuiModalFooter>
            <EuiButtonEmpty onClick={onClose} data-test-subj="editRolloverAliasModalCloseButton">
              Close
            </EuiButtonEmpty>

            <EuiButton onClick={this.onEditRolloverAlias} disabled={!rolloverAlias} fill data-test-subj="editRolloverAliasModalAddButton">
              Edit
            </EuiButton>
          </EuiModalFooter>
        </EuiModal>
      </EuiOverlayMask>
    );
  }
}
