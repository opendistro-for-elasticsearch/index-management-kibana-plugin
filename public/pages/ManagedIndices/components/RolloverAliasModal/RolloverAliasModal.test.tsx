/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent, wait } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import RolloverAliasModal from "./RolloverAliasModal";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";

describe("<RolloverAliasModal /> spec", () => {
  it("renders the component", () => {
    render(<RolloverAliasModal services={browserServicesMock} index="some_index" onClose={() => {}} core={coreServicesMock} />);
    // EuiOverlayMask appends an element to the body so we should have two, an empty div from react-test-library
    // and our EuiOverlayMask element
    expect(document.body.children).toHaveLength(2);
    expect(document.body.children[1]).toMatchSnapshot();
  });

  it("calls close when close button clicked", () => {
    const onClose = jest.fn();
    const { getByTestId } = render(
      <RolloverAliasModal services={browserServicesMock} index="some_index" onClose={onClose} core={coreServicesMock} />
    );

    fireEvent.click(getByTestId("editRolloverAliasModalCloseButton"));
    expect(onClose).toHaveBeenCalled();
  });

  it("disables add button when no alias", async () => {
    const { getByTestId, getByPlaceholderText } = render(
      <RolloverAliasModal services={browserServicesMock} index="some_index" onClose={() => {}} core={coreServicesMock} />
    );

    expect(getByTestId("editRolloverAliasModalAddButton")).toBeDisabled();

    await userEvent.type(getByPlaceholderText("Rollover alias"), "some_alias");

    expect(getByTestId("editRolloverAliasModalAddButton")).not.toBeDisabled();
  });

  it("shows success toaster when successful", async () => {
    browserServicesMock.indexService.editRolloverAlias = jest.fn().mockResolvedValue({ ok: true, response: { acknowledged: true } });
    const { getByTestId, getByPlaceholderText } = render(
      <RolloverAliasModal services={browserServicesMock} index="some_index" onClose={() => {}} core={coreServicesMock} />
    );

    await userEvent.type(getByPlaceholderText("Rollover alias"), "some_alias");

    fireEvent.click(getByTestId("editRolloverAliasModalAddButton"));

    await wait();

    expect(coreServicesMock.notifications.toasts.addSuccess).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addSuccess).toHaveBeenCalledWith("Edited rollover alias on some_index");
  });

  it("shows error toaster when error is thrown", async () => {
    browserServicesMock.indexService.editRolloverAlias = jest.fn().mockRejectedValue(new Error("this is an error"));
    const { getByTestId, getByPlaceholderText } = render(
      <RolloverAliasModal services={browserServicesMock} index="some_index" onClose={() => {}} core={coreServicesMock} />
    );

    await userEvent.type(getByPlaceholderText("Rollover alias"), "some_alias");

    fireEvent.click(getByTestId("editRolloverAliasModalAddButton"));

    await wait();

    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("this is an error");
  });

  it("shows error toaster when error is returned", async () => {
    browserServicesMock.indexService.editRolloverAlias = jest.fn().mockResolvedValue({ ok: false, error: "some error" });
    const { getByTestId, getByPlaceholderText } = render(
      <RolloverAliasModal services={browserServicesMock} index="some_index" onClose={() => {}} core={coreServicesMock} />
    );

    await userEvent.type(getByPlaceholderText("Rollover alias"), "some_alias");

    fireEvent.click(getByTestId("editRolloverAliasModalAddButton"));

    await wait();

    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("some error");
  });

  it("shows error toaster when call is not acknowledged", async () => {
    browserServicesMock.indexService.editRolloverAlias = jest.fn().mockResolvedValue({
      ok: true,
      response: { acknowledged: false },
    });
    const { getByTestId, getByPlaceholderText } = render(
      <RolloverAliasModal services={browserServicesMock} index="some_index" onClose={() => {}} core={coreServicesMock} />
    );

    await userEvent.type(getByPlaceholderText("Rollover alias"), "some_alias");

    fireEvent.click(getByTestId("editRolloverAliasModalAddButton"));

    await wait();

    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("Failed to edit rollover alias on some_index");
  });
});
