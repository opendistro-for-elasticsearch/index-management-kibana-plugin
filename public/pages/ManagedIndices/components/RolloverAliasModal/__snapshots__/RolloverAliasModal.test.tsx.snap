// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RolloverAliasModal /> spec renders the component 1`] = `
<div
  class="euiOverlayMask euiOverlayMask--aboveHeader"
>
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="1"
  />
  <div
    data-focus-lock-disabled="false"
  >
    <div
      class="euiModal euiModal--maxWidth-default"
      tabindex="0"
    >
      <button
        aria-label="Closes this modal window"
        class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
        type="button"
      >
        EuiIconMock
      </button>
      <div
        class="euiModal__flex"
      >
        <div
          class="euiModalHeader"
        >
          <div
            class="euiModalHeader__title"
          >
            Edit rollover alias
          </div>
        </div>
        <div
          class="euiModalBody"
        >
          <div
            class="euiModalBody__overflow"
          >
            <div
              class="euiFormRow"
              id="some_html_id-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="some_html_id"
                >
                  Rollover alias
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFormControlLayout"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      aria-describedby="some_html_id-help"
                      class="euiFieldText"
                      id="some_html_id"
                      placeholder="Rollover alias"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="euiFormHelpText euiFormRow__text"
                  id="some_html_id-help"
                >
                  A rollover alias is required when using the rollover action.
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiModalFooter"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--primary"
            data-test-subj="editRolloverAliasModalCloseButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonEmpty__content"
            >
              <span
                class="euiButtonEmpty__text"
              >
                Close
              </span>
            </span>
          </button>
          <button
            class="euiButton euiButton--primary euiButton--fill euiButton-isDisabled"
            data-test-subj="editRolloverAliasModalAddButton"
            disabled=""
            type="button"
          >
            <span
              class="euiButtonContent euiButton__content"
            >
              <span
                class="euiButton__text"
              >
                Edit
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
</div>
`;
